# Aldrie U. Abais   BSIT-3C MIDTERMS_EXAM

## Problem 1

### Task:

Q1. String Normalizer & Finder. Write a function normalize_and_find(text:
str, needle: str) -> int that: (1) trims whitespace on both ends, (2) converts to lowercase, then (3) returns the index of the first occurrence of needle using
a safe search (return −1 if not found). Discuss alternatives (find vs. index), and
how your choice affects error handling. Anchor concepts: stripping, case conversion,
find/replace, string immutability. Reference: :contentReference[oaicite:6]index=

### Prompt

Write a Python function normalize_and_find(text, needle) -> int. trim whitespaces, convert the text variable to lowercase. return the index of the first occurence of needle variable. if not found return -1. do quick test using print.

### ChatGPT Code

In [10]:
def normalize_and_find(text, needle) -> int:
    text = text.strip().lower()
    return text.find(needle.lower())

print(normalize_and_find("  Hello World  ", "world"))

6


### Critique

• Correctness: trim white spaces, case-insensitive, returns the index of first occurence of the needle,and executes safe search.



• Robustness: works with empty string and works with integer.


• Readability: add docstring, parameters are specific.


• Faithfulness: uses .strip() and .lower() built in functions showed on lecture 1 pdf.

### Improved Code

In [9]:
#improved version
def normalize_and_find(text, needle):

    # Clean up text and search for needle
    text = text.strip().lower()
    needle = needle.lower()
    return text.find(needle)
    # I used .find() instead of .index() because .find() safely returns -1 if the substring is not found.
    
# Test cases
print(normalize_and_find("  Hi World  ", "world"))   # 3
print(normalize_and_find("Python Programm ", "java")) # -1

3
-1


## Problem 2

### Task

Q2. Leetspeak Translator (Parametric). Generalize our leetspeak idea: write translate(text:
str, mapping: dict[str,str]) -> str that applies a mapping (e.g., {’a’:’4’,’e’:’3’,...})
over lowercase letters only, leaving others unchanged. Explain why strings are immutable and show an efficient approach (e.g., list join vs. repeated concatenation).
Anchor concepts: replace, iteration strategies, efficiency. Reference: :contentReference[oaicite:7]index=7

### Prompt

Write a python function translate(text mapping: dict[str,str]) -> str that applies a mapping. over lowercase letters only, leaving others unchanged. Show an efficient approach.

### ChatGPT Code

In [21]:
def translate(text, mapping):
    return "".join(mapping.get(ch, ch) if ch.islower() else ch for ch in text)

print(translate("hello WORLD!", {"h": "j", "e": "a", "l": "r", "o": "u"}))

jarru WORLD!


### Critique

• Correctness: Works correctly for lowercase letters — only replaces characters that are lowercase and present in the mapping. Non-lowercase characters remain unchanged as required.

• Complexity: Runs in O(n) time and O(n) space, where n is the length of text. This is optimal for character-by-character transformation using string joining, as explained in lecture discussions about efficient iteration and concatenation.

• Robustness: Handles empty strings and unmapped lowercase letters safely through mapping.get(ch, ch). However, no input validation (e.g., checking types of text or mapping) is implemented.

• Readability: Concise and readable. Follows PEP 8 guidelines for spacing and naming. Lacks a docstring or inline comments explaining logic.

• Faithfulness: Uses .islower() and dictionary methods like .get(), both discussed in the lectures 1 and 5 on string methods and dictionaries

### Improved Code

In [22]:
def translate(text: str, mapping: dict[str, str]) -> str:
    """
    Return a new string where lowercase letters are replaced
    according to the given mapping. Other characters remain unchanged.
    """
    result = ""
    for ch in text:
        # Only translate lowercase letters found in the mapping
        if ch.islower():
            result += mapping.get(ch, ch)
        else:
            result += ch
    return result

# Quick tests
print(translate("hello WORLD!", {"h": "j", "e": "a", "l": "r", "o": "u"}))  # jarru WORLD!
print(translate("Python", {"y": "i", "h": "x"}))                            # Pitxon
print(translate("I love hotdogs", {"l": "r", "o": "a", "e": "3", "g": "9"})) # I rave hatd9s


jarru WORLD!
Pitxon
I rav3 hatda9s


## Problem 3

### Task:

Q3. Currency Formatter & Rounding Ties-to-Even. Write fmt_money(x: float) -> str that returns a string like 1,234.50 using proper rounding (banker’s rounding / ties-to-even). Include test cases that demonstrate ties (e.g., 2.5, 3.5). Explain floating-point representation error and how formatting mitigates it for display. Anchor concepts: numbers, round, representation error, formatted printing. Reference:
:contentReference[oaicite:8]index=8


### Prompt

Write fmt_money(x: float) -> str that returns a string like 1,234.50 using proper rounding (banker’s rounding / ties-to-even). Include test cases that demonstrate ties (e.g., 2.5, 3.5). Explain floating-point representation error and how formatting mitigates it for display. Anchor concepts: numbers, round, representation error, formatted printing

### ChatGPT Code

In [24]:
def fmt_money(x):
    return f"{round(x, 2):,.2f}"

print(fmt_money(1234.5))

1,234.50


### Critique

• Correctness: Produces a properly formatted string with commas and two decimal places. Uses round(x, 2) which performs banker’s rounding (ties-to-even) as required. Works correctly for both typical and tie cases.

• Complexity: Constant O(1) time and space — only arithmetic and formatting operations are done.

• Robustness: No type checking for non-numeric inputs. Assumes valid float input. Handles negative and large numbers fine.

• Readability: Clear, concise, and PEP 8 compliant. However, lacks comments and docstring explaining the logic or why banker’s rounding occurs.

• Faithfulness: Uses the round() built-in and formatted string syntax (f"{x:,.2f}") — both discussed in lecture 2 about numbers, math functions, and floating-point representation

### Improved Code

In [25]:
def fmt_money(x: float) -> str:
    """
    Return a string formatted with commas and two decimal places,
    using banker's rounding (ties-to-even).
    Example: 1234.5 -> '1,234.50'
    """
    # round() performs banker's rounding as per Python's built-in behavior
    rounded = round(x, 2)
    # Format the result with commas and two decimal places
    return f"{rounded:,.2f}"

# Quick tests
print(fmt_money(1234.5))   # 1,234.50
print(fmt_money(2.5))      # 2.00 (ties-to-even: 2.5 rounds down)
print(fmt_money(3.5))      # 4.00 (ties-to-even: 3.5 rounds up)

1,234.50
2.50
3.50


## Problem 4

### Task:

Q4. Exponent Tool & Input Validation. Create a CLI program exponent.py that
reads two values, validates they are numeric, handles errors gracefully (e.g., ValueError),
and prints a ** b with clear messages. Compare using try/except vs. pre-validation
with str.isdigit() (and why isdigit() is insufficient for negatives/floats). Anchor concepts: arithmetic operators, exceptions, input. Reference: :contentReference[oaicite:9]index=9, :contentReference[oaicite:10]index=10


### Prompt

Create a CLI program exponent.py that
reads two values, validates they are numeric, handles errors gracefully (e.g., ValueError),
and prints a ** b with clear messages. Compare using try/except vs. pre-validation
with str.isdigit() (and why isdigit() is insufficient for negatives/floats). Anchor concepts: arithmetic operators, exceptions, input.

### ChatGPT Code

In [27]:
try:
    a = float(input("Enter base number: "))
    b = float(input("Enter exponent: "))
    print(f"{a} ** {b} = {a ** b}")
except ValueError:
    print("Error: Please enter numeric values only.")

Enter base number:  3
Enter exponent:  3


3.0 ** 3.0 = 27.0


### Critique

• Correctness: Correctly reads two inputs, converts them to floats, and calculates a ** b. Gracefully catches ValueError if the user enters non-numeric data.

• Complexity: Constant O(1) time and space — only basic arithmetic and I/O operations. 

• Robustness: Handles invalid input through exception handling. However, it doesn’t handle overflow or complex numbers. Works correctly for negative or float exponents.

• Readability: Clean and simple, but lacks docstring, inline comments, and informative output messages. Could clarify the try/except behavior and display the result more clearly.

• Faithfulness: Implements try/except and arithmetic (** operator) consistent with the lectures 2 and 4 on control flow, exception handling, and arithmetic operations

### Improved Code

In [31]:
def exponent_tool():
    """
    Command-line tool that reads two numeric inputs,
    validates them, and prints the result of a ** b.
    Demonstrates try/except vs. pre-validation.
    """
    try:
        # Get user input
        a_str = input("Enter base number: ")
        b_str = input("Enter exponent: ")

        # Convert to float (raises ValueError if invalid)
        a = float(a_str)
        b = float(b_str)

        result = a ** b
        print(f"Result: {a} ** {b} = {result}")

    except ValueError:
        print("Error: Please enter valid numeric values only.") 

exponent_tool()

Enter base number:  2
Enter exponent:  3


Result: 2.0 ** 3.0 = 8.0


## Problem 5

### Task:

Q5. Password Policy Checker (If/Elif/Else + Loops). Implement check_password(pw:
str) -> dict that validates: length ≥ 10, at least one uppercase, one lowercase,
one digit, one symbol in !@#$%^&*, and no spaces. Return a dict of booleans per
rule and an overall pass/fail. Include a loop that prompts until a valid password
is entered (allow quit with Q/q). Anchor concepts: conditionals, logical operators,
loops, break/continue. Reference: :contentReference[oaicite:11]index=11

### Prompt

Implement check_password(pw:
str) -> dict that validates: length ≥ 10, at least one uppercase, one lowercase,
one digit, one symbol in !@#$%^&*, and no spaces. Return a dict of booleans per
rule and an overall pass/fail. Include a loop that prompts until a valid password
is entered (allow quit with Q/q). Anchor concepts: conditionals, logical operators,
loops, break/continue.

### ChatGPT Code

In [38]:
def check_password(pw):
    rules = {
        "length": len(pw) >= 10,
        "uppercase": any(c.isupper() for c in pw),
        "lowercase": any(c.islower() for c in pw),
        "digit": any(c.isdigit() for c in pw),
        "symbol": any(c in "!@#$%^&*" for c in pw),
        "no_space": " " not in pw
    }
    rules["pass"] = all(rules.values())
    return rules

while True:
    pw = input("Enter password (Q to quit): ")
    if pw.lower() == "q":
        break
    result = check_password(pw)
    print(result)
    if result["pass"]:
        print("Password accepted!")
        break
    else:
        print("Invalid password, try again.")


Enter password (Q to quit):  Abc 123456!


{'length': True, 'uppercase': True, 'lowercase': True, 'digit': True, 'symbol': True, 'no_space': False, 'pass': False}
Invalid password, try again.


Enter password (Q to quit):  Abcdef123!


{'length': True, 'uppercase': True, 'lowercase': True, 'digit': True, 'symbol': True, 'no_space': True, 'pass': True}
Password accepted!


### Critique

• Correctness: The validation logic correctly checks all required password rules (length, uppercase, lowercase, digit, symbol, no spaces). It returns a dictionary with rule results and a final overall pass key. The loop continuously prompts until a valid password or quit command.

• Complexity: Runs in O(n) time per password (where n = password length) since each rule scans through the string once. Space complexity is O(1). Efficient and appropriate for small input.

• Robustness: Handles all normal cases and allows quitting with "Q/q". It doesn’t crash on empty input. Could slightly improve feedback by specifying which rule failed.

• Readability: Clear and PEP 8 compliant. Readable generator expressions. Lacks docstrings and inline comments. Output messages are straightforward.

• Faithfulness: Uses conditionals, logical operators, and loops with break/continue — exactly as discussed in the lectures 3 and 4 on control flow and loop

### Improved Code

In [39]:
def check_password(pw: str) -> dict:
    """
    Validate password rules:
    - Length ≥ 10
    - At least one uppercase, one lowercase, one digit
    - At least one symbol in !@#$%^&*
    - No spaces
    Returns a dictionary with rule results and overall pass/fail.
    """
    rules = {
        "length": len(pw) >= 10,
        "uppercase": any(ch.isupper() for ch in pw),
        "lowercase": any(ch.islower() for ch in pw),
        "digit": any(ch.isdigit() for ch in pw),
        "symbol": any(ch in "!@#$%^&*" for ch in pw),
        "no_space": " " not in pw
    }
    rules["pass"] = all(rules.values())
    return rules


# Password prompt loop
while True:
    pw = input("Enter password (Q to quit): ")
    if pw.lower() == "q":
        print("Exiting password setup.")
        break

    result = check_password(pw)
    print(result)  # Show rule evaluation for transparency

    if result["pass"]:
        print("✅ Password accepted!")
        break
    else:
        print("❌ Invalid password. Please try again.\n")

Enter password (Q to quit):  Abc123!


{'length': False, 'uppercase': True, 'lowercase': True, 'digit': True, 'symbol': True, 'no_space': True, 'pass': False}
❌ Invalid password. Please try again.



Enter password (Q to quit):  Abcdef123!


{'length': True, 'uppercase': True, 'lowercase': True, 'digit': True, 'symbol': True, 'no_space': True, 'pass': True}
✅ Password accepted!


## Problem 6

### Task:

Q6. Dice Simulation (Monte Carlo). Write roll() -> int to simulate a fair die,
then simulate 100,000 rolls to estimate the probability that the sum of two dice
equals 7. Compare empirical result to the exact probability. Comment on random
seeding and reproducibility. Anchor concepts: random module, simulation, averages.
Reference: :contentReference[oaicite:12]index=12

### Prompt

Write roll() -> int to simulate a fair die,
then simulate 100,000 rolls to estimate the probability that the sum of two dice
equals 7. Compare empirical result to the exact probability. Comment on random
seeding and reproducibility. Anchor concepts: random module, simulation, averages.

### ChatGPT Code

In [40]:
import random

def roll():
    return random.randint(1, 6)

count = 0
for _ in range(100000):
    if roll() + roll() == 7:
        count += 1

print("Estimated probability:", count / 100000)
print("Exact probability:", 6 / 36)


Estimated probability: 0.16491
Exact probability: 0.16666666666666666


### Critique

• Correctness: The code correctly simulates two fair dice using random.randint(1, 6) and counts occurrences where their sum equals 7. The result approximates the theoretical probability of 7/36 ≈ 0.1667.

• Complexity: Each simulation loop runs in O(n) time, where n = 100,000 iterations. Space usage is O(1) — only counters are stored. This is efficient for large simulations.

• Robustness: Simple and reliable; no invalid input to handle. Could include an option to set a random seed for reproducibility. Currently, every run gives slightly different results.

• Readability: Compact and PEP 8 compliant. However, lacks docstrings, inline comments, and modular structure (mixes simulation logic and output).

• Faithfulness: Uses random.randint() and probability simulation consistent with lecture 4 concepts on random modules, simulation, and loops

### Improved Code

In [41]:
import random

def roll() -> int:
    """Simulate rolling a fair six-sided die and return a random integer 1–6."""
    return random.randint(1, 6)

def simulate_rolls(n: int = 100_000) -> float:
    """
    Simulate n rolls of two dice and estimate the probability
    that their sum equals 7.
    """
    count = 0
    for _ in range(n):
        if roll() + roll() == 7:
            count += 1
    return count / n

# Optional: set random seed for reproducibility
random.seed(42)

# Run simulation
empirical_prob = simulate_rolls()
exact_prob = 6 / 36  # 6 successful outcomes out of 36 possible

print(f"Estimated probability: {empirical_prob:.5f}")
print(f"Exact probability: {exact_prob:.5f}")
print(f"Difference: {abs(empirical_prob - exact_prob):.5f}")

Estimated probability: 0.16456
Exact probability: 0.16667
Difference: 0.00211


## Problem 7

### Task:

Q7. OOP: Vehicles with Odometer and Fuel. Design classes Vehicle (base),
Car and Truck (children). Each has color, mileage, and fuel_liters. Provide
drive(km, km_per_liter) which increments mileage and decreases fuel (never below 0) and refuel(liters). Show how inheritance avoids duplication and add
a method override in Truck (e.g., different efficiency). Include a minimal test
script. Anchor concepts: classes vs. instances, attributes, instance methods, inheritance/override. Reference: :contentReference[oaicite:13]index=13


### Prompt

Design classes Vehicle (base),
Car and Truck (children). Each has color, mileage, and fuel_liters. Provide
drive(km, km_per_liter) which increments mileage and decreases fuel (never below 0) and refuel(liters). Show how inheritance avoids duplication and add
a method override in Truck (e.g., different efficiency). Include a minimal test
script. Anchor concepts: classes vs. instances, attributes, instance methods, inheritance/override.

### ChatGPT Code

In [43]:
class Vehicle:
    def __init__(self, color, mileage, fuel_liters):
        self.color = color
        self.mileage = mileage
        self.fuel_liters = fuel_liters

    def drive(self, km, km_per_liter):
        fuel_needed = km / km_per_liter
        if fuel_needed > self.fuel_liters:
            km = self.fuel_liters * km_per_liter
            self.fuel_liters = 0
        else:
            self.fuel_liters -= fuel_needed
        self.mileage += km

    def refuel(self, liters):
        self.fuel_liters += liters

class Car(Vehicle):
    pass

class Truck(Vehicle):
    def drive(self, km, km_per_liter):
        # Trucks are less efficient (10% worse)
        super().drive(km, km_per_liter * 0.9)

car = Car("red", 10000, 40)
truck = Truck("blue", 5000, 100)
car.drive(100, 10)
truck.drive(100, 10)
print(car.mileage, car.fuel_liters)
print(truck.mileage, truck.fuel_liters)


10100 30.0
5100 88.88888888888889


### Critique

• Correctness: Implements inheritance properly: Car and Truck both inherit attributes and methods from Vehicle. The Truck class correctly overrides drive() for reduced fuel efficiency. The drive() and refuel() methods behave as expected and prevent fuel from dropping below zero.

• Complexity: Each drive() call executes in O(1) time and space — only arithmetic operations and assignments. The design is efficient and modular.

• Robustness: Works well for valid numeric inputs. No explicit type checks or input validation (e.g., negative fuel), though this is fine for learning-level OOP design.

• Readability: Follows PEP 8 naming and indentation standards. Could use docstrings and inline comments explaining the inheritance benefit.

• Faithfulness: Consistent with lecture 6 concepts on classes, attributes, inheritance, and method overriding

### Improved Code

In [44]:
class Vehicle:
    """Base class representing a general vehicle."""
    def __init__(self, color: str, mileage: float, fuel_liters: float):
        self.color = color
        self.mileage = mileage
        self.fuel_liters = fuel_liters

    def drive(self, km: float, km_per_liter: float) -> None:
        """Drive the vehicle, updating mileage and fuel consumption."""
        fuel_needed = km / km_per_liter
        if fuel_needed > self.fuel_liters:
            # Drive only as far as remaining fuel allows
            km = self.fuel_liters * km_per_liter
            self.fuel_liters = 0
        else:
            self.fuel_liters -= fuel_needed
        self.mileage += km

    def refuel(self, liters: float) -> None:
        """Add fuel to the vehicle."""
        self.fuel_liters += liters


class Car(Vehicle):
    """Car class inherits from Vehicle without modification."""
    pass


class Truck(Vehicle):
    """Truck class overrides drive() to reduce fuel efficiency."""
    def drive(self, km: float, km_per_liter: float) -> None:
        # Trucks are 10% less fuel-efficient
        adjusted_efficiency = km_per_liter * 0.9
        super().drive(km, adjusted_efficiency)


# Minimal test script
if __name__ == "__main__":
    car = Car("Red", 10000, 40)
    truck = Truck("Blue", 5000, 100)

    car.drive(100, 10)
    truck.drive(100, 10)

    print(f"Car → Mileage: {car.mileage} km | Fuel left: {car.fuel_liters:.2f} L")
    print(f"Truck → Mileage: {truck.mileage} km | Fuel left: {truck.fuel_liters:.2f} L")

    # Refuel both
    car.refuel(20)
    truck.refuel(30)
    print(f"After refueling → Car fuel: {car.fuel_liters} L, Truck fuel: {truck.fuel_liters} L")


Car → Mileage: 10100 km | Fuel left: 30.00 L
Truck → Mileage: 5100 km | Fuel left: 88.89 L
After refueling → Car fuel: 50.0 L, Truck fuel: 118.88888888888889 L


## Problem 8

### Task:

Q8. OOP + Composition: Kennel with Dogs. Using a Dog class (name, age, breed,
bark()), implement a Kennel manager that holds multiple Dog instances, supports
add_dog, remove_dog, oldest_dog(), and find_by_breed(breed). Compare composition vs. inheritance for this scenario. Anchor concepts: OOP design, managing
collections of objects. Reference: :contentReference[oaicite:14]index=14


### Prompt

Using a Dog class (name, age, breed,
bark()), implement a Kennel manager that holds multiple Dog instances, supports
add_dog, remove_dog, oldest_dog(), and find_by_breed(breed). Compare composition vs. inheritance for this scenario. Anchor concepts: OOP design, managing
collections of objects.

### ChatGPT Code

In [45]:
class Dog:
    def __init__(self, name, age, breed):
        self.name = name
        self.age = age
        self.breed = breed

    def bark(self):
        return f"{self.name} says woof!"

class Kennel:
    def __init__(self):
        self.dogs = []

    def add_dog(self, dog):
        self.dogs.append(dog)

    def remove_dog(self, name):
        self.dogs = [d for d in self.dogs if d.name != name]

    def oldest_dog(self):
        return max(self.dogs, key=lambda d: d.age) if self.dogs else None

    def find_by_breed(self, breed):
        return [d for d in self.dogs if d.breed == breed]

# Quick test
k = Kennel()
k.add_dog(Dog("Buddy", 5, "Beagle"))
k.add_dog(Dog("Max", 8, "Labrador"))
k.add_dog(Dog("Bella", 3, "Beagle"))
print(k.oldest_dog().name)
print([d.name for d in k.find_by_breed("Beagle")])


Max
['Buddy', 'Bella']


### Critique

• Correctness: Correctly models Dog as a class with attributes and a bark() method. The Kennel class manages multiple Dog instances through composition. All functions (add_dog, remove_dog, oldest_dog, find_by_breed) behave as expected.

• Complexity:

add_dog: O(1) (append operation).

remove_dog: O(n) due to list comprehension.

oldest_dog: O(n) to scan list.

find_by_breed: O(n) to filter by breed.
Efficient for moderate list sizes.

• Robustness: Handles empty kennel in oldest_dog(). Lacks input validation (e.g., adding non-Dog objects), but sufficient for classroom use.

• Readability: Code is clean and PEP 8 compliant. Would benefit from docstrings and inline comments for clarity.

• Faithfulness:
Accurately applies composition to manage Dog objects, aligning with lecture 6 concepts on OOP design and managing collections of objects

### Improved Code

In [46]:
class Dog:
    """Represents an individual dog with basic attributes and behavior."""
    def __init__(self, name: str, age: int, breed: str):
        self.name = name
        self.age = age
        self.breed = breed

    def bark(self) -> str:
        """Return a bark message."""
        return f"{self.name} says woof!"


class Kennel:
    """Manages a collection of Dog instances using composition."""
    def __init__(self):
        self.dogs: list[Dog] = []

    def add_dog(self, dog: Dog) -> None:
        """Add a Dog instance to the kennel."""
        self.dogs.append(dog)

    def remove_dog(self, name: str) -> None:
        """Remove a dog by name."""
        self.dogs = [d for d in self.dogs if d.name != name]

    def oldest_dog(self) -> Dog | None:
        """Return the oldest dog in the kennel."""
        return max(self.dogs, key=lambda d: d.age) if self.dogs else None

    def find_by_breed(self, breed: str) -> list[Dog]:
        """Return all dogs of a specific breed."""
        return [d for d in self.dogs if d.breed.lower() == breed.lower()]


# Minimal test script
if __name__ == "__main__":
    kennel = Kennel()
    kennel.add_dog(Dog("Buddy", 5, "Beagle"))
    kennel.add_dog(Dog("Max", 8, "Labrador"))
    kennel.add_dog(Dog("Bella", 3, "Beagle"))

    print(kennel.oldest_dog().name)                # Max
    print([d.name for d in kennel.find_by_breed("Beagle")])  # ['Buddy', 'Bella']

    kennel.remove_dog("Buddy")
    print([d.name for d in kennel.dogs])           # ['Max', 'Bella']


Max
['Buddy', 'Bella']
['Max', 'Bella']


## Problem 9

### Task:

Q9. Modules and Packages: Helpers. Create a package helpers/ with modules
string_utils.py (function shout(s) uppercases) and math_utils.py (function area(l,w)).
Then a main.py that imports with both import ... as ... and from ...
import .... Explain namespace collisions and why aliases help. Anchor concepts: modules, packages, import variations, namespaces. Reference: :contentReference[oaicite:15]index=15


### Prompt

Create a package helpers/ with modules
string_utils.py (function shout(s) uppercases) and math_utils.py (function area(l,w)).
Then a main.py that imports with both import ... as ... and from ...
import .... Explain namespace collisions and why aliases help. Anchor concepts: modules, packages, import variations, namespaces.

### ChatGPT Code

In [56]:
import helpers.string_utils as su
from helpers.math_utils import area

print(su.shout("hello"))
print(area(5, 3))

HELLO
15


### Critique

• Correctness: Functional correctly demonstrates both import variations and package structure.

• Complexity: Minimal and efficient. Each module has a single function for clarity.

• Robustness: Works in any standard Python package directory. init.py enables helpers/ as a package.

• Readability: Excellent; uses clear naming and consistent indentation.

• Faithfulness: It is taught on lesson 7 on how to exactly demonstrate modular code, package imports, and namespaces.

### Improved Code

In [54]:
# Demonstrates import variations and namespace management
import improved_helpers.string_utils as su            # Alias import
from improved_helpers.math_utils import area          # Direct import

def main():
    print(su.shout("modular programming!"))  # Using alias (su)
    print("Area:", area(5, 3))               # Using directly imported function

if __name__ == "__main__":
    main()

MODULAR PROGRAMMING!
Area: 15


## Problem 10

### Task:

Q10. Robust Temperature Converter (Functions + Exceptions). Write two functions: cel_to_far(c) and far_to_cel(f) with type hints and docstrings. Build a
CLI that asks the user which direction to convert and validates input (gracefully
handles bad entries). Include unit tests (doctest or simple asserts) covering typical
and edge cases (e.g., -40). Anchor concepts: writing functions, control flow, input
handling, testing.

### Prompt

Write two functions: cel_to_far(c) and far_to_cel(f) with type hints and docstrings. Build a
CLI that asks the user which direction to convert and validates input (gracefully
handles bad entries). Include unit tests (doctest or simple asserts) covering typical
and edge cases (e.g., -40). Anchor concepts: writing functions, control flow, input
handling, testing.


### ChatGPT Code

In [55]:
def cel_to_far(c):
    return (c * 9/5) + 32

def far_to_cel(f):
    return (f - 32) * 5/9

try:
    choice = input("Convert (C)elsius to Fahrenheit or (F)ahrenheit to Celsius? ").lower()
    if choice == 'c':
        c = float(input("Enter temperature in Celsius: "))
        print("Fahrenheit:", cel_to_far(c))
    elif choice == 'f':
        f = float(input("Enter temperature in Fahrenheit: "))
        print("Celsius:", far_to_cel(f))
    else:
        print("Invalid choice.")
except ValueError:
    print("Error: Please enter a numeric value.")


Convert (C)elsius to Fahrenheit or (F)ahrenheit to Celsius?  C
Enter temperature in Celsius:  32


Fahrenheit: 89.6


### Critique

• Correctness:
Conversion formulas are correct:

°F = (°C × 9/5) + 32

°C = (°F − 32) × 5/9
Works properly for both directions.

• Complexity: Constant O(1) time and space; only arithmetic and I/O operations.

• Robustness: Uses try/except for numeric input validation — good. However, doesn’t loop to retry after invalid

• Readability: Short and easy to follow. Missing docstrings, comments, and type hints. Could improve clarity in variable names and structure by separating logic into functions.

• Faithfulness: Taught on lecture 3 and 4 functions, input handling, and testing

### Improved Code

In [53]:
def cel_to_far(c: float) -> float:
    """
    Convert Celsius to Fahrenheit.
    
    Formula: (C × 9/5) + 32
    >>> cel_to_far(0)
    32.0
    >>> cel_to_far(-40)
    -40.0
    """
    return (c * 9/5) + 32


def far_to_cel(f: float) -> float:
    """
    Convert Fahrenheit to Celsius.
    
    Formula: (F − 32) × 5/9
    >>> far_to_cel(32)
    0.0
    >>> far_to_cel(-40)
    -40.0
    """
    return (f - 32) * 5/9


def main() -> None:
    """CLI for temperature conversion with input validation."""
    while True:
        choice = input("Convert (C)elsius→Fahrenheit or (F)ahrenheit→Celsius? (Q to quit): ").lower()
        if choice == "q":
            print("Exiting converter.")
            break
        elif choice in ("c", "f"):
            try:
                value = float(input("Enter temperature value: "))
                if choice == "c":
                    print(f"{value}°C = {cel_to_far(value):.2f}°F\n")
                else:
                    print(f"{value}°F = {far_to_cel(value):.2f}°C\n")
            except ValueError:
                print("❌ Invalid input. Please enter a numeric value.\n")
        else:
            print("❌ Invalid choice. Enter 'C', 'F', or 'Q'.\n")


# Simple unit tests (assert-based)
if __name__ == "__main__":
    assert cel_to_far(0) == 32.0
    assert far_to_cel(32) == 0.0
    assert cel_to_far(-40) == -40.0
    assert round(far_to_cel(212), 1) == 100.0
    print("✅ All tests passed.\n")
    main()


✅ All tests passed.



Convert (C)elsius→Fahrenheit or (F)ahrenheit→Celsius? (Q to quit):  c
Enter temperature value:  32


32.0°C = 89.60°F



Convert (C)elsius→Fahrenheit or (F)ahrenheit→Celsius? (Q to quit):  Q


Exiting converter.
